### PR TITLE
fix(storage): wrap error when MaxAttempts is hit

### PR DESCRIPTION
**Pull request by tritone**
Wrap the error when retries are cut off by hitting the configured value for MaxAttempts. This makes it easier to verify that retries occurred.

Also adds emulator tests verifying that wrapping occurs as expected for timeout errors and MaxAttempts.

Updates #9720

### CODE REVIEW

File: storage/client_test.go [hunk 1]
```diff
@@ -1351,31 +1352,57 @@ func TestObjectConditionsEmulated(t *testing.T) {
...
+		testID := createRetryTest(t, project, bucket, client, instructions)
+		ctx = callctx.SetHeaders(ctx, "x-retry-test-id", testID)
+		_, err := client.GetBucket(ctx, bucket, nil, withRetryConfig(&retryConfig{policy: RetryNever}))
 
-		attrs, err := client.CreateBucket(ctx, project, bucket, &BucketAttrs{}, nil)
-		if err != nil {
-			t.Fatalf("creating bucket: %v", err)
+		var ae *apierror.APIError
+		if errors.As(err, &ae) {
+			// We espect a 503/UNAVAILABLE error. For anything else including a nil
```
Comment (BrennaEpp) [inline storage/client_test.go:1362]:
```suggestion
			// We expect a 503/UNAVAILABLE error. For anything else including a nil
```

File: storage/client_test.go [hunk 2]
```diff
@@ -1351,31 +1352,57 @@ func TestObjectConditionsEmulated(t *testing.T) {
...
+		instructions := map[string][]string{"storage.buckets.get": {"return-503", "return-503", "return-503", "return-503", "return-503"}}
+		testID := createRetryTest(t, project, bucket, client, instructions)
+		ctx = callctx.SetHeaders(ctx, "x-retry-test-id", testID)
+		ctx, cancel := context.WithTimeout(ctx, 100*time.Millisecond)
+		defer cancel()
+		_, err := client.GetBucket(ctx, bucket, nil, idempotent(true))
+
+		var ae *apierror.APIError
+		if errors.As(err, &ae) {
+			// We espect a 503/UNAVAILABLE error. For anything else including a nil
```
Comment (BrennaEpp) [inline storage/client_test.go:1384]:
```suggestion
			// We expect a 503/UNAVAILABLE error. For anything else including a nil
```

File: storage/client_test.go [hunk 3]
```diff
@@ -1385,9 +1412,44 @@ func TestRetryNeverEmulated(t *testing.T) {
 				t.Errorf("GetBucket: got unexpected error %v; want 503", err)
 			}
 		}
+		// Error should be wrapped so it indicates that MaxAttempts has been reached.
+		if got, want := err.Error(), "retry failed after 3 attempts"; !strings.Contains(got, want) {
+			t.Errorf("got error: %q, want to contain: %q", got, want)
+		}
```
Comment (BrennaEpp) [inline storage/client_test.go:1418]: Do we need to delete the retry test resource in the testbench after finishing?

**Review by BrennaEpp — COMMENTED** [review]
Awesome!

File: storage/client_test.go (hunk 3 above)
Comment (tritone) [reply storage/client_test.go:1418]: I don't think it matters too much given that the emulator is torn down between test runs, but on the other hand I might as well -- will update.

**Review by BrennaEpp — APPROVED** [review]

**Pull request merged**